File: MyProject/Content/MyAsset/Lua/MyLua/Libs/EventHandle/AddOnceAndCallOnceEventDispatch.lua
MLoader("MyLua.Libs.Core.GlobalNS");
MLoader("MyLua.Libs.Core.Class");
MLoader("MyLua.Libs.EventHandle.EventDispatch");

local M = GlobalNS.Class(GlobalNS.EventDispatch);
M.clsName = "AddOnceAndCallOnceEventDispatch";
GlobalNS[M.clsName] = M;

function M:ctor()
    
end

function M:addEventHandle(pThis, handle)
    if (not self:isExistEventHandle(pThis, handle)) then
        M.super.addEventHandle(self, pThis, handle);
    end
end

function M:dispatchEvent(dispatchObject)
    M.super.dispatchEvent(self, dispatchObject);
    self:clearEventHandle();
end

return M;

File: Project/MyProject/MyProject/Source/MyProject/Private/BasicFrame/Libs/EventHandle/AddOnceAndCallOnceEventDispatch.cpp
#include "MyProject.h"
#include "AddOnceAndCallOnceEventDispatch.h"

MY_BEGIN_NAMESPACE(MyNS)

void AddOnceAndCallOnceEventDispatch::addEventHandle(EventDispatchDelegate handle)
{
	if (!this->isExistEventHandle(handle))
	{
		Super::addEventHandle(handle);
	}
}

void AddOnceAndCallOnceEventDispatch::dispatchEvent(IDispatchObject* dispatchObject)
{
	Super::dispatchEvent(dispatchObject);
	this->clearEventHandle();
}

MY_END_NAMESPACE